dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,407 @@
+
+These are another features which are taken into account (soft requirements).
+
+| Feature                  | Description                                                               |
+| ------------------------ | ------------------------------------------------------------------------- |
+| Microkernel              | Microkernels are more secure and are preferred                            |
+| OS portability           | Required to avoid rebuilding minimal OS for each device                   |
+| CPU Architecture support | OS supported architectures, mostly we consider x86, ARM, RISC-V and POWER |
+
+### OS score counting
+
+The table at the bottom of this document contains summary of features of all
+OSes together with theirs scores. Each score is computed using the following
+rules
+
+- If feature is present then +1
+- If feature is not present then 0, or if feature would be hard to implement
+  then -1
+- If this is a hard requirement then multiple result by 2
+
+## Different OSs propositions
+
+The research effect is presented below. 4 systems were considered:
+
+* Zephyr
+* Xous
+* seL4
+* Linux
+
+Each of them has a short description, an analysis of the launch in DLME, and the
+possibilities and potential problems that will have to be addressed for the
+Fobnail Token to be functional.
+
+### Zephyr
+
+Zephyr is a well-known and supported embedded RTOS with a monolithic kernel
+written in C. It supports various devices ranging from Cortex-M to x86 CPUs.
+It's main benefits are IPv4, CoAP, and USB (including USB EEM) support which are
+required for communication with Fobnail token. Other benefits are listed
+[here](https://www.zephyrproject.org/benefits/).
+
+#### Running in DLME
+
+Secure launch uses a component called
+[SKL](https://github.com/TrenchBoot/secure-kernel-loader) (Secure Kernel Loader)
+which is responsible for booting OS in DLME (Zephyr in that case). Zephyr boots
+using Multiboot1 but SKL does not support it (only Multiboot2 is supported). For
+SKL to be able to boot Zephyr, either Zephyr would have to support Multiboot2 or
+SKL Multiboot1.
+
+Also, Zephyr may have trouble with running on different hardware configurations.
+Some important HW-related configuration is baked into Zephyr during build:
+
+- LAPIC base is selected by `CONFIG_LOAPIC_BASE_ADDRESS`, if target platform has
+  LAPIC remapped using `IA32_APIC_BASE` Zephyr will break.
+
+- APIC mode (xAPIC vs x2APIC) is selected when building Zephyr and cannot be
+  detected at runtime. Zephyr may not boot if it has different APIC mode set
+  than firmware has.
+
+#### Fobnail integration
+
+Zephyr provides most of the drivers needed for integration:
+
+- USB hub drivers
+- USB EEM driver
+- Network stack with IPv4 and CoAP support
+
+TPM driver is missing, however there is a fairly recent
+[PoC implementation](https://github.com/drandreas/zephyr-tpm2-poc) of TPM2
```

Comment:
AFAICT SPI only, for embedded devices, which may not help with implementing it on TPM exposed as compatible with PC Client specification. Still better than nothing, I guess...

Reply:
@krystian-hebel I added a note

Reply:
@krystian-hebel can we resolve this thread?